# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/MacroAssembler.cpp
# Commit: d6506b5ee07b
# Full Hash: d6506b5ee07b5c65f889ce7c2214b0172f205898
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-16 09:43:03
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r 80a164c1054e -r d6506b5ee07b js/src/jit/MacroAssembler.cpp
--- a/js/src/jit/MacroAssembler.cpp	Wed Sep 15 11:02:57 2021 +0000
+++ b/js/src/jit/MacroAssembler.cpp	Wed Sep 15 11:10:15 2021 +0000
@@ -3941,22 +3941,13 @@
     addPtr(index, scratch);
   }
 
-  storePtr(WasmTlsReg,
-           Address(getStackPointer(), WasmCallerTLSOffsetBeforeCall));
-  loadPtr(Address(scratch, offsetof(wasm::FunctionTableElem, tls)), WasmTlsReg);
-  storePtr(WasmTlsReg,
-           Address(getStackPointer(), WasmCalleeTLSOffsetBeforeCall));
+  loadPtr(Address(scratch, offsetof(wasm::FunctionTableElem, code)), scratch);
 
   Label nonNull;
-  branchTest32(Assembler::NonZero, WasmTlsReg, WasmTlsReg, &nonNull);
+  branchTest32(Assembler::NonZero, scratch, scratch, &nonNull);
   wasmTrap(wasm::Trap::IndirectCallToNull, trapOffset);
   bind(&nonNull);
 
-  loadWasmPinnedRegsFromTls();
-  switchToWasmTlsRealm(index, WasmTableCallScratchReg1);
-
-  loadPtr(Address(scratch, offsetof(wasm::FunctionTableElem, code)), scratch);
-
   return call(desc, scratch);
 }
 